src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -0,0 +1,177 @@
+                const { lexeme } = token;
+                remainingInputString = remainingInputString.substring(lexeme.length);
+                tokens.push(token);
+            } else {
+                // If no tokens has been matched, mark current character as unknown token.
+                // If the previous token is also unknown, append current character to
+                // the previous token.
+                const prevToken = tokens[tokens.length - 1];
+                if (prevToken && prevToken.type === TokenType.unknown) {
+                    prevToken.lexeme += remainingInputString.substring(0, 1);
+                } else {
+                    tokens.push({
+                        type: TokenType.unknown,
+                        lexeme: remainingInputString.substring(0, 1),
+                    });
+                }
+                remainingInputString = remainingInputString.substring(1);
+            }
+        }
+        return tokens;
+    }
+
+    public extractWordsAndPhrases(inputString: string): string[] {
+        return (
+            this.tokenize(inputString)
+                // Filter only words and phrases
+                .filter((token) => token.type === TokenType.word || token.type === TokenType.phrase)
+                .map((wordOrPhrase) => {
+                    // `phrase` tokens require some dirty preprocessing
+                    // to remove quotes and possible inner wildcards.
+                    if (wordOrPhrase.type === TokenType.phrase) {
+                        return wordOrPhrase.lexeme.replace(/('|"|\*)/g, '').trim();
+                    }
+                    return wordOrPhrase.lexeme;
+                })
+                // After `phrase` tokens preprocessing, some strings can become empty,
+                // so we should remove them.
+                .filter((str) => str)
+        );
+    }
+
+    /**
+     * `tokenizeCurrentChunk` tries to find the best possible match for the current
+     * part of the input string
+     */
+    private static tokenizeCurrentChunk(remainingInputString: string): Token | undefined {
+        let tokenCandidate: Token | undefined;
+        BooleanQueryTokenizer.tokenPatterns.forEach((tokenPattern) => {
+            const [name, tokenRegExp] = tokenPattern;
+            const lexemeMatch = remainingInputString.match(tokenRegExp);
+            if (lexemeMatch !== null) {
+                const [lexeme] = lexemeMatch;
+                if (!tokenCandidate) {
+                    tokenCandidate = { type: name, lexeme };
+                }
+                // Longer lexeme always wins. If tokens' lexemes are equal in length, then wins a
+                // token who's lexeme has been matched by a pattern with higher priority.
+                // Which means the current one (because we iterate through patterns from left to
+                // right and right patterns have higher priority).
+                if (lexeme.length >= tokenCandidate.lexeme.length) {
```

Comment:
This condition will still be true also when the candidate was just set for the first time, I would still use `else if` here. It will also be clearer to read the logic (even though it is quite clear given the comment 👍 )

Reply:
That's true, addressed